include/swift/AST/DiagnosticsClangImporter.def

```diff
@@ -228,7 +228,18 @@ ERROR(foreign_reference_types_invalid_retain,none,
 ERROR(foreign_reference_types_invalid_release,none,
       "specified release function '%0' is invalid; release function must have exactly "
       "one argument of type '%1'", (StringRef, StringRef))
-
+ERROR(foreign_reference_types_invalid_retain_return_type, none,
+      "specified retain function '%0' is invalid; retain function must have "
+      "'void' return type",
+      (StringRef))
+ERROR(foreign_reference_types_invalid_release_return_type, none,
+      "specified release function '%0' is invalid; release function must have "
+      "'void' return type",
+      (StringRef))
+ERROR(foreign_reference_types_invalid_retain_not_a_function_decl, none,
+      "specified retain function '%0' is not a function", (StringRef))
+ERROR(foreign_reference_types_invalid_release_not_a_function_decl, none,
```

Comment:
You could potentially cut the diagnostics in half by doing something like `%select{retain|release}0` and pass an additional bool parameter to the diagnostic.